File: Python/Labs/Module 1/Lab 2/Q1 - Valid Sudoku/main.py
def v(b,num):
    def r(b,num):
        for i in b:
            n=set()
            for c in i:
                if c!=".":
                    if c in n:
                        return False
                    n.add(c)
        return True
    def c(b,num):
        for i in range(num):
            n=set()
            for j in range(num):
                c=b[j][i]
                if c!='.':
                    if c in n:
                        return False
                    n.add(c)
        return True
    def x(b,num):
        for i in range(0,9,3):
            for j in range(0,9,3):
                n=set()
                for k in range(i,i+3):
                    for l in range(j,j+3):
                        c=b[k][l]
                        if c!='.':
                            if c in n:
                                return False
                            n.add(c)
        return True
    return r(b,num) and c(b,num) and x(b,num)
n=int(input())
b=[]
for _ in range(n):
    r=input().split()
    b.append(r)
vld=v(b,n)
if vld==True:
    print("YES")
else:
    print("NO")
                                                                                                                            

File: Java/Labs/Module 1/Lab 2/Q1 - Valid Sudoku/Main.java
import java.util.*;

public class Main {

    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        String x=s.nextLine();
        int[][] b=new int[9][9];
        for(int i=0;i<9;i++){
            String[] r=s.nextLine().split(" ");
            for(int j=0;j<9;j++){
                if(r[j].equals(".")){
                    b[i][j]=0;
                } else{
                    b[i][j]=Integer.parseInt(r[j]);
                }
            }
        }
        System.out.println(v(b));
        s.close();
    }
    public static String v(int[][] b){
        for(int i=0;i<9;i++){
            if(!c(b,i,0,i,8)){
                return "NO";
            }
        }
        for(int j=0;j<9;j++){
            if(!c(b,0,j,8,j)){
                return "NO";
            }
        }
        for(int r=0;r<9;r+=3){
            for(int c=0;c<9;c+=3){
                if(!c(b,r,c,r+2,c+2)){
                    return "NO";
                }
            }
        }
        return "YES";
    }
    public static boolean c(int[][] b,int sr,int sc, int er,int ec){
        boolean[] v=new boolean[10];
        for(int i=sr;i<=er;i++){
            for(int j=sc;j<=ec;j++){
                int n=b[i][j];
                if(n!=0){
                    if(v[n]){
                        return false;
                    }
                    v[n]=true;
                }
            }
        }
        return true;
    }
}
                                                                                                                            